dstep/translator/Context.d

```diff
@@ -198,9 +219,72 @@ bool shouldBeAnonymous(Context context, Cursor cursor)
 }
 
 /**
-  * Returns true, if cursor is in the global scope.
-  */
+ * Returns true, if cursor is in the global scope.
+ */
 bool isGlobal(Cursor cursor)
 {
     return cursor.semanticParent.kind == CXCursorKind.CXCursor_TranslationUnit;
 }
+
+/**
+ * The collectTypeNames function scans the whole AST of the translation unit and produces
+ * a set of the type names in global scope.
+ *
+ * The type names are required for parsing C code (e.g. macro definition bodies),
+ * as C grammar isn't context free.
+ */
+
+bool[string] collectTypeNames(TranslationUnit translUnit)
+{
+    void collectBasicTypes(ref bool[string] result)
+    {
+        result["void"] = true;
+        result["char"] = true;
+        result["signed char"] = true;
+        result["unsigned char"] = true;
+        result["short"] = true;
+        result["signed short"] = true;
+        result["unsigned short"] = true;
+        result["signed"] = true;
+        result["unsigned"] = true;
+        result["int"] = true;
+        result["signed int"] = true;
+        result["unsigned int"] = true;
+        result["long"] = true;
+        result["signed long"] = true;
+        result["unsigned long"] = true;
+        result["long long"] = true;
+        result["signed long long"] = true;
+        result["unsigned long long"] = true;
+        result["float"] = true;
+        result["double"] = true;
+    }
```

Comment:
* It seems that (un)signed can be specified in any position, before, after and in between (in the case of long long)
* `short int`, `long int`, `long double` are also valid types
* `const` can be specified for any type
* `__complex__` and `_Complex` in any combination of the floating point types are valid as well. The floating point type does not need to be specified
* `_Bool` and `bool` (if `<stdbool.h>` is included) are valid types in C99. I tried compiling code containing these types with `-std=c89 -Wall` and got no errors or warnings